### PR TITLE
Remove @NotNull from the codebase

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
The `@NotNull` annotation is applied sporadically throughout the codebase, providing little value.  This PR removes it. Objects that may be null should be properly annotated with `@Nullable`, and other objects should be implicitly treated as `@NotNull`. I believe there have been prior discussions favoring moving away from `@NotNull`.

To avoid being a breaking change for the time being, this PR deprecates various `NonNullXYZ` utility classes provided by forge that were largely used by the old capability system. These would otherwise just be removed.

This PR also adds the following:
1. Automatic generation of `package-info.java` files via the `generatePackageInfos` task. [generate-package-infos.gradle](https://github.com/neoforged/NeoForge/pull/553/files#diff-29cbe77c4ba4075976af2185cfd96db048506fce4e417fa4c76fadaab757c6ed)
2. Deduplication of spotless rules to a single location. [spotless-rules.gradle](https://github.com/neoforged/NeoForge/pull/553/files#diff-81abdc9a43b4bd538d66ca3c44fe36175a1afb3329e3b57e21dd5c9b44ac5369)
3. The `applyAllFormatting` task which runs `generatePackageInfos`, `licenseFormat`, and `spotlessApply`. [apply-all-formatting.gradle](https://github.com/neoforged/NeoForge/pull/553/files#diff-d17f70a932125c84bda991e2db2754471f274cb301d965436702f0ce301baba3)

### CODE REVIEW

**Comment by Gaming32** [comment]
You could use `ParametersAreNonnullByDefault`, `FieldsAreNonnullByDefault`, and `MethodsReturnNonnullByDefault` on the package level as a replacement, as Minecraft already does.

**Review by Technici4n — CHANGES_REQUESTED** [review]
Big fan of this change!

Can you add a spotless rule against `@NotNull` as well? (See the hacky custom steps that we have, both in the main gradle and the subprojects).

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]
just removing `NutNull` is not a good idea, please add the default annotations (`ParametersAreNonnullByDefault`, `FieldsAreNonnullByDefault` and `MethodsReturnNonnullByDefault`) to the packages

**Comment by Matyrobbrt** [comment]
We should make a task to generate package infos as FAPI does.

File: src/main/java/net/neoforged/neoforge/common/util/NonNullConsumer.java [hunk 1]
```diff
@@ -6,14 +6,14 @@
...
-import org.jetbrains.annotations.NotNull;
 
 /**
  * Equivalent to {@link Consumer}, except with nonnull contract.
  *
  * @see Consumer
+ * @deprecated Use {@link Consumer}
  */
 @FunctionalInterface
 public interface NonNullConsumer<T> {
```
Comment (PlatinPython) [inline src/main/java/net/neoforged/neoforge/common/util/NonNullConsumer.java]: Missing a `@Deprecated` annotation here

**Review by PlatinPython — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/common/data/SoundDefinitionsProvider.java [hunk 2]
```diff
@@ -207,7 +207,7 @@ private boolean validate(final String name, final SoundDefinition.Sound sound) {
         // Differently from all the other errors, this is not a 'missing sound' but rather something completely different
         // that has broken the invariants of this sound definition's provider. In fact, a sound may only be either of
         // SOUND or EVENT type. Any other values is somebody messing with the internals, reflectively adding something
-        // to an enum or passing `null` to a parameter annotated with `@NotNull`.
+        // to an enum or passing `null` to a parameter annotated with ``.
```
Comment (Mari023) [inline src/main/java/net/neoforged/neoforge/common/data/SoundDefinitionsProvider.java]: this comment need some rewording now

**Comment by Shadows-of-Fire** [comment]
Autogen of `package-info.java` added, and comments addressed.  See updated main post.

File: build.gradle [hunk 3]
```diff
@@ -106,3 +90,7 @@ spotless {
         bumpThisNumberIfACustomStepChanges(2)
     }
 }
+
+apply from: rootProject.file('buildscript/spotless-rules.gradle')
```
Comment (Minecraftschurli) [inline build.gradle:94]: shouldn't at least part of the 'patches' format move here too

File: buildscript/generate-package-infos.gradle [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                    |import javax.annotation.ParametersAreNonnullByDefault;
+                    |import net.minecraft.FieldsAreNonnullByDefault;
+                    |import net.minecraft.MethodsReturnNonnullByDefault;
+                """.stripMargin().trim()
+
+                packageInfoFile.text = pkgInfoText
+            }
+        }
+    }
+}
```
Comment (Minecraftschurli) [inline buildscript/generate-package-infos.gradle:27]: I suggest making this an actual task class with action and not a default task with doLast

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 3 above)
Comment (Shadows-of-Fire) [reply build.gradle:94]: Probably not, this was isolating the part that was shared between `build.gradle`, `tests/build.gradle`, and `testframework/build.gradle` - the patches formatting only appears in `build.gradle`, and I'm not sure if there would be side effects of it existing in the subprojects

File: buildscript/generate-package-infos.gradle (hunk 4 above)
Comment (Shadows-of-Fire) [reply buildscript/generate-package-infos.gradle:27]: That can be saved for someone with better gradle/groovy knowledge, I'm already burnt out from getting to this point.

File: buildscript/generate-package-infos.gradle [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import java.util.regex.Matcher
+
+def generatePackageInfos = tasks.register('generatePackageInfos', Task) {
```
Comment (Technici4n) [inline buildscript/generate-package-infos.gradle:3]: This could be in NG. (Probably should be)

Comment (Shadows-of-Fire) [reply buildscript/generate-package-infos.gradle:3]: tasks for a later time™

Comment (Minecraftschurli) [reply buildscript/generate-package-infos.gradle:3]: more like GU, I don't think it fits into NG

**Review by Technici4n — APPROVED** [review]

**Review by marchermans — APPROVED** [review]

**Review by Minecraftschurli — APPROVED** [review]

**Comment by Tslat** [comment]
Don't forget to update everything that should be null pls

image

**Comment by Shadows-of-Fire** [comment]
Yeah that's going to be a much harder lint to accomplish.  Should probably open an issue to track violations.

**Pull request merged**